Source/WarpX.cpp

```diff
@@ -993,10 +993,47 @@ WarpX::ReadParameters ()
 
         if (maxLevel() > 0) {
             Vector<Real> lo, hi;
-            utils::parser::getArrWithParser(pp_warpx, "fine_tag_lo", lo);
-            utils::parser::getArrWithParser(pp_warpx, "fine_tag_hi", hi);
-            fine_tag_lo = RealVect{lo};
-            fine_tag_hi = RealVect{hi};
+            bool fine_tag_lo_specified = utils::parser::queryArrWithParser(pp_warpx, "fine_tag_lo", lo);
+            bool fine_tag_hi_specified = utils::parser::queryArrWithParser(pp_warpx, "fine_tag_hi", hi);
+            std::string ref_patch_function;
+            bool parser_specified = pp_warpx.query("ref_patch_function(x,y,z)",ref_patch_function);
+            const std::string error_string = std::string( "For max_level > 0, you need to either set\
+                                             warpx.fine_tag_lo and warpx.fine_tag_hi\
+                                             or warpx.ref_patch_function(x,y,z)");
+            WARPX_ALWAYS_ASSERT_WITH_MESSAGE( ((fine_tag_lo_specified && fine_tag_hi_specified) ||
+                                                parser_specified ),
+                                                error_string);
```

Comment:
Elsewhere, the standard has been to put the error message directly in the call to `WARPX_ALWAYS_ASSERT_WITH_MESSAGE`, instead of defining a temporary string.

Reply:
I see! 
I saw this code for reference https://github.com/ECP-WarpX/WarpX/blob/ce82a694bcf0bdcb425ad0ad73c9658430ea9663/Source/Diagnostics/BoundaryScrapingDiagnostics.cpp#L54 

Would be nice to document the standard practice for writing error messages?

Reply:
I used "standard" loosely. I was just noticing what was done elsewhere, but I don't think there is any official standard. I agree that it would be a good idea to have some list of standard practices for details such as this.

Reply:
sounds good.  Either way, I have removed the temp string :)